src/hmm/hsp.py

```diff
@@ -0,0 +1,194 @@
+
+        Will always return false if the object compared is not an instance of HSP
+
+        Args:
+            __o (object): Target object to compare to
+
+        Returns:
+            bool: True if __o and self are equal
+        """
+        if not isinstance(__o, HSP):
+            return False
+
+        return all(
+            [
+                __o.cds.nt_start == self.cds.nt_start,
+                __o.cds.nt_stop == self.cds.nt_stop,
+                __o.domain == self.domain,
+                __o.score == self.score,
+            ]
+        )
+
+    @staticmethod
+    def has_overlap(hsp_a: HSP, hsp_b: HSP) -> bool:
+        """Return True if there is overlap between this
+
+        Args:
+            cds_b (CDS): CDS to compare
+
+        Returns:
+            bool: whether there is overlap between this cds and another
+        """
+        has_overlap = HSP.len_overlap(hsp_a, hsp_b) > 0
+        return has_overlap
+
+    @staticmethod
+    def len_overlap(hsp_a: HSP, hsp_b: HSP) -> int:
+        """Return the length of the overlap between this CDS and another
+
+        Args:
+            cds_b (CDS): CDS to compare
+
+        Returns:
+            int: length of the overlap between this CDS and another
+        """
+
+        if hsp_a.env_start < hsp_b.env_start:
+            left = hsp_b.env_start
+        else:
+            left = hsp_a.env_start
+
+        if hsp_a.env_stop > hsp_b.env_stop:
+            right = hsp_b.env_stop
+        else:
+            right = hsp_a.env_stop
+
+        # limit to > 0
+        return max(0, right - left)
+
+    @staticmethod
+    def overlap_filter(hsp_list: list[HSP], overlap_cutoff=0.1) -> list[HSP]:
+        """Filters overlapping HSP
+
+        Args:
+            hsp_list (list[HSP]): a list of high scoring protein hits
+            overlap_cutoff: The maximum percentage sequence length domains can overlap
+                without being discarded
+
+        Returns:
+            list[HSP]: a filtered list of high scoring protein hits
+        """
+        # for this the hsps have to be sorted by any of the start coordinates
+        # this is for the rare case where two hsps have the same bit score.
+        # to replicate BiG-SCAPE output, the hsp with a lower start coordinate
+        # will end up being chosen
+        sorted_hsps = sorted(hsp_list, key=lambda hsp: hsp.cds.nt_start)
+
+        new_list = []
+        for i in range(len(sorted_hsps) - 1):
+            for j in range(i + 1, len(sorted_hsps)):
+                hsp_a: HSP = sorted_hsps[i]
+                hsp_b: HSP = sorted_hsps[j]
+
+                # check if we are the same CDS
+                if hsp_a == hsp_b:
+                    # check if there is overlap between the domains
+                    if not HSP.has_overlap(hsp_a, hsp_b):
+                        continue
+
+                    len_aa_overlap = HSP.len_overlap(hsp_a, hsp_b)
+
+                    overlap_perc_loc1 = len_aa_overlap / (
```

Comment:
whats loc? same as hsp? add comment " check what percentage of domains is in overlap" or along these lines